src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -0,0 +1,676 @@
+
+		while ( $p->next_tag( array( 'tag_closers' => 'visit' ) ) && false === $unbalanced ) {
+			$tag_name = $p->get_tag();
+
+			if ( 'SVG' === $tag_name || 'MATH' === $tag_name ) {
+				$unbalanced = true;
+				break;
+			}
+
+			if ( $p->is_tag_closer() ) {
+				list( $opening_tag_name, $directives_prefixes ) = end( $tag_stack );
+
+				if ( 0 === count( $tag_stack ) || $opening_tag_name !== $tag_name ) {
+
+					/*
+					 * If the tag stack is empty or the matching opening tag is not the
+					 * same than the closing tag, it means the HTML is unbalanced and it
+					 * stops processing it.
+					 */
+					$unbalanced = true;
+					break;
+				} else {
+					// Remove the last tag from the stack.
+					array_pop( $tag_stack );
+
+					/*
+					 * If the matching opening tag didn't have any directives, it can skip
+					 * the processing.
+					 */
+					if ( 0 === count( $directives_prefixes ) ) {
+						continue;
+					}
+				}
+			} else {
+				$directives_prefixes = array();
+
+				foreach ( $p->get_attribute_names_with_prefix( 'data-wp-' ) as $attribute_name ) {
+
+					/*
+					 * Extracts the directive prefix to see if there is a server directive
+					 * processor registered for that directive.
+					 */
+					list( $directive_prefix ) = $this->extract_prefix_and_suffix( $attribute_name );
+					if ( array_key_exists( $directive_prefix, self::$directive_processors ) ) {
+						$directives_prefixes[] = $directive_prefix;
+					}
+				}
+
+				/*
+				 * If this tag will visit its closer tag, it adds it to the tag stack
+				 * so it can process its closing tag and check for unbalanced tags.
+				 */
+				if ( $p->has_and_visits_its_closer_tag() ) {
+					$tag_stack[] = array( $tag_name, $directives_prefixes );
+				}
+			}
+
+			/*
+			 * Sorts the attributes by the order of the `directives_processor` array
+			 * and checks what directives are present in this element. The processing
+			 * order is reversed for tag closers.
+			 */
+			$directives_prefixes = array_intersect(
+				$p->is_tag_closer()
+					? $directive_processor_prefixes_reversed
+					: $directive_processor_prefixes,
+				$directives_prefixes
+			);
+
+			// Executes the directive processors present in this element.
+			foreach ( $directives_prefixes as $directive_prefix ) {
+				$func = is_array( self::$directive_processors[ $directive_prefix ] )
+					? self::$directive_processors[ $directive_prefix ]
+					: array( $this, self::$directive_processors[ $directive_prefix ] );
+				call_user_func_array(
+					$func,
+					array( $p, &$context_stack, &$namespace_stack )
+				);
+			}
+		}
+
+		/*
+		 * It returns the original content if the HTML is unbalanced because
+		 * unbalanced HTML is not safe to process. In that case, the Interactivity
+		 * API runtime will update the HTML on the client side during the hydration.
+		 */
+		return $unbalanced || 0 < count( $tag_stack ) ? $html : $p->get_updated_html();
```

Comment:
Should there be some warning emitted to the console that client-side hydration is happening?

Reply:
Sure 🙂

I'm not familiar with the WP APIs to do that. Do you have any suggestions?

Reply:
Two options I see.

1. This method could use `wp_trigger_error()` which will output a notice to the error log. However, devs may not see this. I've essentially [done this](https://github.com/WordPress/performance/blob/801b657736e9820b47b81ae3d69ab0ef43fefe70/modules/images/image-loading-optimization/class-ilo-html-tag-processor.php#L247-L260) in my Image Loading Optimization plugin.
2. Some HTML attribute could be added to the block markup like the following and client-side JS could look for those attributes to then log them to the console:

```php
if ( $unbalanced || 0 < count( $tag_stack ) ) {
	$p = new WP_HTML_Tag_Processor( $html );
	if ( $p->next_tag() ) {
		$p->set_attribute( 'data-wp-error', __( 'Failed to hydrate block due to unbalanced HTML.' ) );
		return $p->get_updated_html();
	}
}
return $p->get_updated_html();
```

Reply:
Ok, thanks, Weston. I'll think about how to display an error in the console.

I've added it to the [Epic](https://github.com/WordPress/gutenberg/issues/56803) (under debug) and added it here as a follow-up.